src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -1069,8 +1069,153 @@ theorem rotateLeft_eq_rotateLeftAux_of_lt {x : BitVec w} {r : Nat} (hr : r < w)
     x.rotateLeft r = x.rotateLeftAux r := by
   simp only [rotateLeft, Nat.mod_eq_of_lt hr]
 
+
+/-- Accessing bits in `x.rotateLeft r` the range `[0, r)` is equal to
+    accesing bits `x` in the range `[w - r, w)`,
+    given by `i : [0, r) ↦ w - r + i : [w - r, w)`.
+
+  Proof by example:
+    Let x := <6 5 4 3 2 1 0> : BitVec 7. See that `x[i] = i` in this visualization.
+  x.rotateLeft 2
+    = (<6 5 | 4 3 2 1 0>).rotateLeft 2
+    = <3 2 1 0 | 6 5>
+
+  (x.rotateLeft 2).getLsb ⟨i, i < 2⟩
+    = <3 2 1 0 | 6 5>.getLsb ⟨i, i < 2⟩
+    = <6 5>[i]
+    = <6 5 | 4 3 2 1 0>[i + len(<4 3 2 1 0>)]
+    = <6 5 | 4 3 2 1 0>[i + 7 - 2]
+
+    Intuitively, grab the full width (7), then move the marker `|` by `r` to the right `(-2)`
+    Then, access the bit at `i` from the right `(+i)`.
+ -/
+theorem getLsb_rotateLeftAux_of_le {x : BitVec w} {r : Nat} {i : Nat} (hi : i < r) :
+    (x.rotateLeftAux r).getLsb i = x.getLsb (w - r + i) := by
+  rw [rotateLeftAux, getLsb_or, getLsb_ushiftRight]
+  suffices (x <<< r).getLsb i = false by
+    simp only [getLsb_shiftLeft, Bool.or_iff_right_iff_imp, Bool.and_eq_true, decide_eq_true_eq,
+      Bool.not_eq_true', decide_eq_false_iff_not, Nat.not_lt, and_imp]
+    omega
+  simp only [getLsb_shiftLeft, Bool.and_eq_false_imp, Bool.and_eq_true, decide_eq_true_eq,
+    Bool.not_eq_true', decide_eq_false_iff_not, Nat.not_lt, and_imp]
+  omega
+
+/-- Accessing bits in `x.rotateLeft r` the range `[r, w)` is equal to
+    accesing bits `x` in the range `[0, w - r)`,
+    given by `i : [r, w) ↦ i - r : [0, w - r)`.
+
+  Proof by example:
+    Let x := <6 5 4 3 2 1 0> : BitVec 7. See that `x[i] = i` in this visualization.
+  x.rotateLeft 2
+    = (<6 5 | 4 3 2 1 0>).rotateLeft 2
+    = <3 2 1 0 | 6 5>
+
+  (x.rotateLeft 2).getLsb ⟨i, i ≥ 2⟩
+    = <3 2 1 0 | 6 5>.getLsb ⟨i, i ≥ 2⟩
+    = <3 2 1 0>[i - 2]
+    = <6 5 | 3 2 1 0>[i - 2]
+
+    Intuitively, grab the full width (7), then move the marker `|` by `r` to the right `(-2)`
+    Then, access the bit at `i` from the right `(+i)`.
+ -/
+theorem getLsb_rotateLeftAux_of_geq {x : BitVec w} {r : Nat} {i : Nat} (hi : i ≥ r) :
+    (x.rotateLeftAux r).getLsb i = (decide (i < w) && x.getLsb (i - r)) := by
+  rw [rotateLeftAux, getLsb_or]
+  suffices (x >>> (w - r)).getLsb i = false by
+    rw [this]
+    simp only [getLsb_shiftLeft, Bool.or_false, hi]
+    have hiltr : decide (i < r) = false := by
+      simp [hi]
+    rw [hiltr]
+    simp
```

Comment:
You can golf this proof by folding the various rewrite steps into a single `simp`. I'll leave it up to your sense of aesthetics as to whether this is actually better.
```suggestion
    have hiltr : decide (i < r) = false := by
      simp [hi]
    simp [getLsb_shiftLeft, Bool.or_false, hi, hiltr, this]
```